diff --git a/CMakeLists.txt b/CMakeLists.txt
index 90822a3..2e00044 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -23,6 +23,9 @@ if(NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
     set(CMAKE_BUILD_TYPE "Release" CACHE STRING "Choose the type of build." FORCE)
 endif()
 
+set(CMAKE_CXX_FLAGS "$CMAKE_CXX_FLAGS -g -O0")
+set(CMAKE_C_FLAGS "$CMAKE_C_FLAGS -g -O0")
+
 # set the module path, used for includes
 set(CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}/cmake")
 
